pkg/cmd/roachtest/operations/changefeeds/utils.go

```diff
@@ -0,0 +1,384 @@
+func getJobIDStateAndHighWM(
+	ctx context.Context, conn *gosql.DB, jobID string,
+) (state jobs.State, highWaterTimestamp hlc.Timestamp, err error) {
+	// Construct SQL query to get job state by job ID.
+	statement := fmt.Sprintf("SELECT status, COALESCE(high_water_timestamp, '0') FROM [SHOW CHANGEFEED JOBS] WHERE job_id='%s';", jobID)
+
+	// Execute query and extract the state value.
+	err = helpers.ExecuteQuery(ctx, func(rowValues []string) error {
+		state = jobs.State(rowValues[0])
+		highWaterTimestamp, err = hlc.ParseHLC(rowValues[1])
+		return err
+	}, conn, statement)
+
+	return state, highWaterTimestamp, err
+}
+
+// changeStateOrCreateChangefeed executes the specified action (e.g., PAUSE, CANCEL) on a randomly
+// selected changefeed job from the list of jobs in currentStates.
+// It confirms the success by checking if the job transitions to one of the expectedFinalStates.
+// If no existing jobs are found it attempts to create a changefeed.
+// It uses the checkAndCreateChangefeed function to determine if a new changefeed can be created.
+func changeStateOrCreateChangefeed(
+	ctx context.Context,
+	o operation.Operation,
+	c cluster.Cluster,
+	action string,
+	currentStates, expectedFinalStates []jobs.State,
+) error {
+	// Connect to the cluster and ensure the connection is closed after use.
+	conn := c.Conn(ctx, o.L(), 1, option.VirtualClusterName(roachtestflags.VirtualCluster))
+	defer func() {
+		_ = conn.Close()
+	}()
+	// Fetch all changefeed jobs with the given current states.
+	cfJobs, err := fetchAllChangefeedJobsWithStates(ctx, conn, currentStates...)
+	if err != nil {
+		return err
+	}
+	if len(cfJobs) == 0 {
+		o.Status(fmt.Sprintf("No jobs found for %s. attempting create...", action))
+		err = checkAndCreateChangefeed(ctx, o, conn, true)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Initialize a random number generator to select a job randomly.
+	r, _ := randutil.NewPseudoRand()
+
+	// Randomly pick a job to perform the action on.
+	jobIndex := randutil.RandIntInRange(r, 0, len(cfJobs))
+	jobToAction := cfJobs[jobIndex]
+
+	// Execute the action (e.g., PAUSE or CANCEL) on the chosen job.
+	o.Status(fmt.Sprintf("Executing %s on job %s", action, jobToAction.jobID))
+	_, err = conn.ExecContext(ctx, fmt.Sprintf("%s JOB %s;", action, jobToAction.jobID))
+	if err != nil {
+		return err
+	}
+
+	// Wait for the job to transition to one of the expected states (e.g., paused, canceled).
+	err = waitForJobToComplete(ctx, conn, action, jobToAction, expectedFinalStates)
+	if err != nil {
+		return err
+	}
+
+	// Log the successful completion of the action.
+	o.Status(fmt.Sprintf("%s complete for Job %s", action, jobToAction.jobID))
+	return nil
+}
+
+// waitForJobToComplete polls the job's state until it transitions to one of the expectedFinalStates
+// (e.g., paused, canceled), or until a timeout occurs.
+func waitForJobToComplete(
```

Comment:
rename: waitForJobState, and update comments, since you are using it to wait for running jobs too